source/common/network/dns_impl.cc

```diff
@@ -52,25 +52,41 @@ void DnsResolverImpl::PendingResolution::onAresHostCallback(int status, hostent*
     delete this;
     return;
   }
+  if (status == ARES_SUCCESS || !fallback_if_failed) {
+    completed_ = true;
+  }
+
   std::list<Address::InstanceConstSharedPtr> address_list;
-  completed_ = true;
   if (status == ARES_SUCCESS) {
-    ASSERT(hostent->h_addrtype == AF_INET);
-    for (int i = 0; hostent->h_addr_list[i] != nullptr; ++i) {
-      ASSERT(hostent->h_length == sizeof(in_addr));
-      sockaddr_in address;
-      memset(&address, 0, sizeof(address));
-      // TODO(mattklein123): IPv6 support.
-      address.sin_family = AF_INET;
-      address.sin_port = 0;
-      address.sin_addr = *reinterpret_cast<in_addr*>(hostent->h_addr_list[i]);
-      address_list.emplace_back(new Address::Ipv4Instance(&address));
+    if (hostent->h_addrtype == AF_INET) {
+      for (int i = 0; hostent->h_addr_list[i] != nullptr; ++i) {
+        ASSERT(hostent->h_length == sizeof(in_addr));
+        sockaddr_in address;
+        memset(&address, 0, sizeof(address));
+        address.sin_family = AF_INET;
+        address.sin_port = 0;
+        address.sin_addr = *reinterpret_cast<in_addr*>(hostent->h_addr_list[i]);
+        address_list.emplace_back(new Address::Ipv4Instance(&address));
+      }
+    } else if (hostent->h_addrtype == AF_INET6) {
+      for (int i = 0; hostent->h_addr_list[i] != nullptr; ++i) {
+        ASSERT(hostent->h_length == sizeof(in6_addr));
+        sockaddr_in6 address;
+        memset(&address, 0, sizeof(address));
+        address.sin6_family = AF_INET6;
+        address.sin6_port = 0;
+        address.sin6_addr = *reinterpret_cast<in6_addr*>(hostent->h_addr_list[i]);
+        address_list.emplace_back(new Address::Ipv6Instance(address));
+      }
     }
+  } else if (fallback_if_failed) {
+    fallback_if_failed = false;
+    getHostByName(AF_INET);
```

Comment:
I think there's a dangerous corner case here. If the first `getHostByName()` defers to asynch completion but then the fallback `getHostByName()` succeeds immediately (synchronously), it will invoke the callback, which will `delete this;`. This will then return to the caller method belonging to a now deleted object.